tools/mtouch/BuildTasks.mtouch.cs

```diff
@@ -73,160 +79,153 @@ protected int Start ()
+		protected override void CompilationFailed (int exitCode)
 		{
-			foreach (var abi in abis)
-				Create (tasks, abi, target, ifile);
+			throw ErrorHelper.CreateError (5103, "Failed to compile the file(s) '{0}'. Please file a bug report at http://bugzilla.xamarin.com", string.Join ("', '", CompilerFlags.SourceFiles.ToArray ()));
 		}
+	}
 
-		public static void Create (List<BuildTask> tasks, Abi abi, Target target, string ifile)
+	class PinvokesTask : CompileTask
+	{
+		protected override void CompilationFailed (int exitCode)
 		{
-			var arch = abi.AsArchString ();
-			var ext = target.App.FastDev ? ".dylib" : ".o";
-			var ofile = Path.Combine (target.App.Cache.Location, "lib" + Path.GetFileNameWithoutExtension (ifile) + "." + arch + ext);
+			throw ErrorHelper.CreateError (4002, "Failed to compile the generated code for P/Invoke methods. Please file a bug report at http://bugzilla.xamarin.com");
+		}
+	}
 
-			if (!Application.IsUptodate (ifile, ofile)) {
-				var task = new PinvokesTask ()
-				{
-					Target = target,
-					Abi = abi,
-					InputFile = ifile,
-					OutputFile = ofile,
-					SharedLibrary = target.App.FastDev,
-					Language = "objective-c++",
-				};
-				if (target.App.FastDev) {
-					task.InstallName = "lib" + Path.GetFileNameWithoutExtension (ifile) + ext;
-					task.CompilerFlags.AddFramework ("Foundation");
-					task.CompilerFlags.LinkWithXamarin ();
-				}
-				tasks.Add (task);
-			} else {
-				Driver.Log (3, "Target '{0}' is up-to-date.", ofile);
+	class RunRegistrarTask : BuildTask
+	{
+		public Target Target;
+		public string RegistrarM;
+		public string RegistrarH;
+
+		public override IEnumerable<string> Inputs {
+			get {
+				foreach (var asm in Target.Assemblies)
+					yield return asm.FullPath;
 			}
+		}
 
-			target.LinkWith (ofile);
-			target.LinkWithAndShip (ofile);
+		public override IEnumerable<string> Outputs {
+			get {
+				yield return RegistrarH;
+				yield return RegistrarM;
+			}
 		}
 
-		protected override void Build ()
+		protected override void Execute ()
 		{
-			if (Compile () != 0)
-				throw new MonoTouchException (4002, true, "Failed to compile the generated code for P/Invoke methods. Please file a bug report at http://bugzilla.xamarin.com");
+			Target.StaticRegistrar.Generate (Target.Assemblies.Select ((a) => a.AssemblyDefinition), RegistrarH, RegistrarM);
 		}
 	}
 
-	internal class RegistrarTask : CompileTask
+	class CompileRegistrarTask : CompileTask
 	{
-		public static void Create (List<BuildTask> tasks, IEnumerable<Abi> abis, Target target, string ifile)
-		{
-			foreach (var abi in abis)
-				Create (tasks, abi, target, ifile);
-		}
-
-		public static void Create (List<BuildTask> tasks, Abi abi, Target target, string ifile)
-		{
-			var app = target.App;
-			var arch = abi.AsArchString ();
-			var ofile = Path.Combine (app.Cache.Location, Path.GetFileNameWithoutExtension (ifile) + "." + arch + ".o");
+		public string RegistrarM;
+		public string RegistrarH;
```

Comment:
I figured that although these are ugly, they're succinct and easy to understand. But what about `RegistrarCodePath` and `RegistrarHeaderPath`?